shutdown.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+)
+
+// Shutdowner provides a method that can manually trigger the shutdown of the
+// application by sending a signal to all open Done channels. Shutdowner works
+// both on applications using Run as well as Start, Done, and Stop.
+type Shutdowner interface {
+	Shutdown(...ShutdownOption) error
+}
+
+// Shutdown broadcasts a signal to all of the Done channels on the application
+// thereby enabling the start of the Stop process.
+func (s *shutdowner) Shutdown(opts ...ShutdownOption) error {
+	return s.broadcastSignal(syscall.SIGTERM)
```

Comment:
nit: ordering: entity generally, we go for: type declaration, constructor (if any), methods.

```
type shutdowner struct {
  ..
}

func (s *shutdowner) Shutdown(...) error {
  ..
}
```

(Except when the methods of a single type are spread across multiple files, so
this doesn't apply to `app` methods below.)

Reply:
I was wondering if there was a particular pattern — thanks!